core/src/main/scala/org/apache/spark/metrics/ExecutorMetricType.scala

```diff
@@ -59,6 +60,43 @@ case object JVMOffHeapMemory extends ExecutorMetricType {
   }
 }
 
+case object ProcessTreeJVMRSSMemory extends ExecutorMetricType {
+  override private[spark] def getMetricValue(memoryManager: MemoryManager): Long = {
+    ExecutorMetricType.pTreeInfo.updateAllMetrics()
```

Comment:
I still don't like how this is actually updating all the other metrics -- it makes the code more confusing to follow, as you have to know there is a relationship between all of the metrics.  I understand that you want to do the work once and grab all the metrics, but we should find a better way to do that.  I see how the current api makes that hard to do.

I have two suggestions:

1) change the api to have `getMetricValue()` also pass in the System.currentTimeInMillis().  Then every single metric type would compare the passed in time against the last time you computed the metrics -- if it was stale, it would recompute everything and update the time.

2) Change the api to allow one "metric getter" object to actually supply multiple metrics.  You'd have a simple implementation which woudl just provide one metric, and you'd change all the existing metrics to extend that simple case, but your implementation would provide multiple metrics in one go.

I actually think (2) is better (that is what I did in memory-monitor plugin) though its a bit more work.  You might need to play with this a bit.

thoughts @edwinalu @mccheah ?

Reply:
I think it makes sense for the metrics provider API to return a `Map[String, Long]` for a set of "named" metrics - we've talked before about attaching a schema to the metrics bundle passed around by this API. So, similar to option 2 above.

Reply:
Should make ProcessTreeMemory extends ExecutorMetricType and individual metrics can be returned from it. This also makes the assumption of calculating the metrics only in the ProcessTreeJVMRSSMemory and subsequent calls using it. We shouldn't depend on the order here.

Reply:
Sorry for the delayed response. Thanks for adding the total memory metrics -- these will be very useful. Agreed that doing the work once is better, but that having ProcessTreeJVMRSSMemory.getMetricValue() update all the metrics is confusing, especially if a user at some point wants to call getMetricValue() for one of the other metrics, and not ProcessTreeJVMRSSMemory.

@squito 's #1 is probably the easiest to make the change for with the existing code. However, #2 with @mccheah's suggestion to return Map sounds best/cleanest as an API, with @dhruve 's suggestion to consolidate into ProcessTreeMemory -- I prefer this approach as well.

Right now the call for getMetricValue is done in Heartbeater.getCurrentMetrics(), and it's mapping ExecutorMetricType.values to the array of actual values. Translating the returned maps to an array (with index mapping to name rather than ExecutorMetricType) will involve some more code. In retrospect, getting the current metrics is probably better done by ExecutorMetrics iteself, rather than having Heartbeater exposed to the implementation details -- would you be able to move the logic there?

Reply:
I changed this to use a map instead of an array of metrics to implement #2 of what @squito suggested